server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalVariableWidthHistogram.java

```diff
@@ -0,0 +1,467 @@
+        return new Bucket(key.doubleValue(), new Bucket.BucketBounds(key.doubleValue(), key.doubleValue()),
+            docCount, format, aggregations);
+    }
+
+    @Override
+    public List<Bucket> getBuckets() {
+        return Collections.unmodifiableList(buckets);
+    }
+
+    @Override
+    public Number getKey(MultiBucketsAggregation.Bucket bucket) {
+        return ((Bucket) bucket).centroid;
+    }
+
+    @Override
+    public Number nextKey(Number key) {
+        return nextKey(key.doubleValue());
+    }
+
+    /**
+     * This method should not be called for this specific subclass of InternalHistogram, since there should not be
+     * empty buckets when clustering.
+=    */
+    private double nextKey(double key){ return key + 1; }
+
+    private static class IteratorAndCurrent {
+
+        private final Iterator<Bucket> iterator;
+        private Bucket current;
+
+        IteratorAndCurrent(Iterator<Bucket> iterator) {
+            this.iterator = iterator;
+            current = iterator.next();
+        }
+
+    }
+
+    private List<Bucket> reduceBuckets(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        // Use a min heap (priority queue) to merge the lists of buckets together
+        final PriorityQueue<IteratorAndCurrent> pq = new PriorityQueue<IteratorAndCurrent>(aggregations.size()) {
+            @Override
+            protected boolean lessThan(IteratorAndCurrent a, IteratorAndCurrent b) {
+                return Double.compare(a.current.centroid, b.current.centroid) < 0;
+            }
+        };
+        for (InternalAggregation aggregation : aggregations) {
+            InternalVariableWidthHistogram histogram = (InternalVariableWidthHistogram) aggregation;
+            if (histogram.buckets.isEmpty() == false) {
+                pq.add(new IteratorAndCurrent(histogram.buckets.iterator()));
+            }
+        }
+
+        List<Bucket> reducedBuckets = new ArrayList<>();
+        while (pq.size() > 0) {
+            final IteratorAndCurrent top = pq.top();
+            reducedBuckets.add(top.current);
+
+            if (reduceContext.isFinalReduce() == false) {
+                reduceContext.consumeBucketsAndMaybeBreak(1);
+            } else {
+                reduceContext.consumeBucketsAndMaybeBreak(-countInnerBucket(top.current));
+            }
+
+            if (top.iterator.hasNext()) {
+                final Bucket next = top.iterator.next();
+                assert next.compareKey(top.current) >= 0 : "shards must return data sorted by key";
+                top.current = next;
+                pq.updateTop();
+            } else {
+                pq.pop();
+            }
+        }
+
+        mergeClosestBuckets(reducedBuckets, numClusters, reduceContext); // Reduce to <code>numClusters</code> buckets
+
+        orderBucketsByXContentKey(reducedBuckets, reduceContext);
+
+        return reducedBuckets;
```

Comment:
Sounds good. I updated it so that overlapping buckets are only merged on the final reduce. It turns out that `InternalAutoDateHistogram` works the same way.

Although, I’m not sure if there is anything we would want to do on partial reduce (`InternalAutoDateHistogram` doesn’t do anything then). Any thoughts?

Reply:
`scripted_metric` does *nothing* on final reduce. It makes me sad! It looks like `auto_date_histogram` does at least a little more. To be honest I'd be happy to get *something* in that doesn't do much on partial reduce and iterate on it. I don't really know what the right thing to do there is yet.